{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9e0e2f50-abb8-4a01-9569-01bcdc73509a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pyspark\n",
    "from pyspark.sql import SparkSession\n",
    "from pyspark.sql.functions import col, explode\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import os\n",
    "import sys\n",
    "from pathlib import Path\n",
    "os.environ['R_HOME'] = sys.exec_prefix+\"/lib/R/\"\n",
    "\n",
    "project_directory = '/Cranio_Lab/Louk_Seton/mesenchyme_project_2023'\n",
    "os.chdir(os.path.expanduser(\"~\")+project_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1eeb519c-1f4f-4647-994b-88b73025cbe4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### Download opentargets and embl database\n",
    "# !wget --directory-prefix=GWAS/opentargets/variant_index/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/genetics/22.09/lut/variant-index/\n",
    "# !wget --directory-prefix=GWAS/opentargets/v2g_scored/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/genetics/22.09/v2g_scored/\n",
    "# !wget --directory-prefix=GWAS/opentargets/v2g/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/genetics/22.09/v2g/\n",
    "# !wget --output-document=GWAS/Catalog/gwas_catalog_v1.0-associations.tsv https://www.ebi.ac.uk/gwas/api/search/downloads/full\n",
    "# !wget --output-document=GWAS/Catalog/gwas_catalog_v1.0.2-associations.tsv https://www.ebi.ac.uk/gwas/api/search/downloads/alternative\n",
    "# !wget --directory-prefix=GWAS/opentargets/platform/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/24.06/output/etl/parquet/associationByOverallDirect\n",
    "# !wget --directory-prefix=GWAS/opentargets/platform/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/24.06/output/etl/parquet/associationByOverallIndirect\n",
    "# !wget --directory-prefix=GWAS/opentargets/platform/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/24.06/output/etl/parquet/diseases\n",
    "# !wget --directory-prefix=GWAS/opentargets/platform/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/24.06/output/etl/parquet/associationByDatatypeDirect\n",
    "# !wget --directory-prefix=GWAS/opentargets/platform/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/24.06/output/etl/parquet/associationByDatatypeIndirect\n",
    "# !wget --directory-prefix=GWAS/opentargets/platform/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/24.06/output/etl/parquet/associationByDatasourceIndirect\n",
    "# !wget --directory-prefix=GWAS/opentargets/platform/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/24.06/output/etl/parquet/associationByDatasourceDirect\n",
    "# !wget --directory-prefix=GWAS/opentargets/platform/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/24.06/output/etl/parquet/targets\n",
    "# !wget --directory-prefix=GWAS/opentargets/platform/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/platform/24.06/output/etl/parquet/evidence/sourceId=ot_genetics_portal/\n",
    "#!wget --directory-prefix=GWAS/opentargets/platform/ --recursive --no-parent --no-host-directories --cut-dirs 8 ftp://ftp.ebi.ac.uk/pub/databases/opentargets/genetics/latest/v2g_scored/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "593064d9-6ffe-486f-95de-17ea1263c851",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyspark\n",
    "from pyspark.sql import SparkSession\n",
    "from pyspark.sql.functions import col, explode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "241a22cb-f5c2-4251-b3a0-9131bd72ae8d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "25/08/13 10:05:17 WARN Utils: Your hostname, cranio213 resolves to a loopback address: 127.0.1.1; using 172.16.4.213 instead (on interface enp1s0f0)\n",
      "25/08/13 10:05:17 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address\n",
      "Setting default log level to \"WARN\".\n",
      "To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).\n",
      "25/08/13 10:05:17 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable\n"
     ]
    }
   ],
   "source": [
    "spark = SparkSession.builder \\\n",
    "           .getOrCreate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "eeb5b050-1af7-434a-a8e6-a94a1971a2b4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ot_genetics = \"GWAS/opentargets/\"\n",
    "ot_platform = \"GWAS/opentargets/platform/\"\n",
    "variant_index = spark.read.parquet(ot_genetics+\"variant_index/\")\n",
    "v2g_scored = spark.read.parquet(ot_genetics+\"v2g_scored/\")\n",
    "v2g = spark.read.parquet(ot_genetics+\"v2g/\")\n",
    "associationByOverallDirect = spark.read.parquet(ot_platform+\"associationByOverallDirect/\")\n",
    "associationByOverallIndirect = spark.read.parquet(ot_platform+\"associationByOverallIndirect/\")\n",
    "associationByDatatypeDirect = spark.read.parquet(ot_platform+\"associationByDatatypeDirect/\")\n",
    "associationByDatatypeIndirect = spark.read.parquet(ot_platform+\"associationByDatatypeIndirect/\")\n",
    "associationByDatasourceIndirect = spark.read.parquet(ot_platform+\"associationByDatasourceIndirect/\")\n",
    "associationByDatasourceDirect = spark.read.parquet(ot_platform+\"associationByDatasourceDirect/\")\n",
    "evidence_ot_genetics_portal = spark.read.parquet(ot_platform+\"evidence/sourceId=ot_genetics_portal/\")\n",
    "diseases = spark.read.parquet(ot_platform+\"diseases/\")\n",
    "targets = spark.read.parquet(ot_platform+\"targets/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9c9d388a-4056-46b4-be6a-c003dc858653",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "evidence_ot_genetics_portal = spark.read.parquet(ot_platform+\"evidence/sourceId=ot_genetics_portal/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "1f83e4d2-c69f-4a15-a829-35b742d1876c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "629"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "associationByOverallIndirect.persist()\n",
    "#count total\n",
    "(associationByOverallIndirect\n",
    "      .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "      .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "      .dropDuplicates(['targetId'])\n",
    "    .count())\n",
    "\n",
    "#keep all\n",
    "list((associationByOverallIndirect\n",
    "      .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "      .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "      .dropDuplicates(['targetId']) #drop duplicates\n",
    "      .orderBy('score', ascending = False) #?order again\n",
    "      .select('targetId')\n",
    "      .toPandas()\n",
    ")['targetId'])\n",
    "\n",
    "#filter by score\n",
    "list((associationByOverallIndirect\n",
    "      .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "      .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "      .dropDuplicates(['targetId']) #drop duplicates\n",
    "      .filter(col(\"score\")>0.4)\n",
    "      .orderBy('score', ascending = False) #?order again\n",
    "      .select('targetId')\n",
    "      .toPandas()      \n",
    ")['targetId'])\n",
    "\n",
    "#filter by top n genes\n",
    "list((associationByOverallIndirect\n",
    "      .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "      .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "      .dropDuplicates(['targetId']) #drop duplicates\n",
    "      .orderBy('score', ascending = False) #?order again\n",
    "      .limit(10)\n",
    "      .select('targetId')\n",
    "      .toPandas()      \n",
    ")['targetId'])\n",
    "\n",
    "associationByOverallIndirect.unpersist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f9128747-105b-4283-ad88-f0feb42dea6a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#gwas query function\n",
    "#Supply a disease code, check the database for associated genes. Cutoff either by score or by top n genes. Can use overall score, limit to results from specific datasource, \n",
    "#use database score. List of genes is human, needs to be moved to mouse ortholog.\n",
    "\n",
    "###code step by step\n",
    "\n",
    "#done - check if OT database is present, if not error\n",
    "#done - Check if disease code exists in database, if not report error\n",
    "#done - check if datasource exists (if not using overall score), if not report error\n",
    "    #either direct or indirect\n",
    "#done - Include disease children yes/no ###be sure to include original term as well\n",
    "    #if yes, report the codes of the children\n",
    "#done - Sort by score of datasource of choice\n",
    "#done - drop duplicates in correct order\n",
    "#done - report total unique genes found\n",
    "#compare with mouse, find orthologs\n",
    "    #report number of orthologs found\n",
    "    #if we want the gene name, change ensemblid to gene name\n",
    "    #########################\n",
    "    #this part needs some work, need to properly drop NA only in the ID column and replace missing gene names with ID. Now we might toss genes that do map.\n",
    "    #need to write a function to retrieve ensembl orthologs by itself\n",
    "    #need to write a case for human genes to skip orthologs but still need to map them to gene names if wanted\n",
    "\n",
    "#Filter by selected method, (all, score cut-off (between 0-1), top n genes)\n",
    "#return list of genes (either by ensembl id or gene name)\n",
    "\n",
    "############\n",
    "#once finished I need to rewrite the code and take out some of the subfunctions for readability and easy understanding\n",
    "#should create a class with all the required subfunctions\n",
    "\n",
    "\n",
    "#### import libraries ####\n",
    "import os\n",
    "import pyspark\n",
    "from pyspark.sql import SparkSession\n",
    "from pyspark.sql.functions import col, explode\n",
    "from typing import Union\n",
    "\n",
    "def OT_genes_by_disease(\n",
    "    disease_ID: str = 'MONDO_0016064', #disease id, set default to cleft palate\n",
    "    gene_cutoff: Union[int, float, str] = 'All', #All for all genes, cutoff score between 0-1, or top n genes cutoff\n",
    "    datasource: list = ['overall'], #somehow check if datasource exists, default uses overall. list of datasources\n",
    "    evidence_type: str = 'indirect', #whether to use direct or indirect evidence\n",
    "    database_location: str = 'GWAS/opentargets/platform/', #location of the database\n",
    "    include_descendants: bool = True, #whether or not to include disease descendants\n",
    "    ortholog_map: str = 'GWAS/ensembl_human_orthologues/Human_to_Mouse.txt', #Location of csv file mapping human genes to focal species. Expects columns in the following order, Human gene stable ID, Human gene name, Focal species stable ID, focal species gene name.\n",
    "    return_gene_names: bool = True, #whether to return gene names or ensemblid\n",
    "\n",
    "    \n",
    "):   \n",
    "    #some checks on variables\n",
    "    datasource_list = ['overall','europepmc', 'crispr', 'uniprot_variants', 'genomics_england', 'gene_burden', 'chembl', 'slapenrich', 'crispr_screen', \n",
    "                       'expression_atlas','intogen', 'reactome', 'orphanet', 'sysbio', 'cancer_biomarkers', 'cancer_gene_census', 'progeny', 'clingen',\n",
    "                       'eva', 'gene2phenotype', 'uniprot_literature','eva_somatic', 'ot_genetics_portal', 'impc']\n",
    "    if datasource not in datasource_list:\n",
    "        if not set(datasource).issubset(datasource_list):\n",
    "            raise ValueError(\"Datasource does not exist. Make sure datasource is 'overall' or a valid datasource.\")\n",
    "    \n",
    "    evidence_type_list = ['direct', 'indirect']\n",
    "    if evidence_type not in evidence_type_list:\n",
    "        raise ValueError(\"evidence_type needs to be either 'direct' or 'indirect'.\")\n",
    "    \n",
    "    #test if database directory exists\n",
    "    if not os.path.isdir(database_location):\n",
    "        raise ValueError('Database directory does not exist')\n",
    "    \n",
    "    #start spark session\n",
    "    spark = SparkSession.builder \\\n",
    "            .getOrCreate()\n",
    "    \n",
    "    #load database\n",
    "    try:\n",
    "        diseases = spark.read.parquet(database_location+\"diseases/\") #load diseases reference\n",
    "        diseases.persist() #persist diseases for quicker loading\n",
    "        #now we want to load the correct database\n",
    "        datasource_filter = False\n",
    "        if evidence_type=='indirect': #if we want indirect evidence\n",
    "            if datasource==['overall']: #if we want to look at overall scores\n",
    "                spark_association_database = spark.read.parquet(database_location+\"associationByOverallIndirect/\")\n",
    "            else: #if we want to look at other datasource scores\n",
    "                datasource_filter = True                \n",
    "                spark_association_database_pre = spark.read.parquet(database_location+\"associationByDatasourceIndirect/\") #name it pre so we can filter afterwards and use the same code\n",
    "        elif evidence_type=='direct':#if we want direct evidence\n",
    "            if datasource==['overall']: \n",
    "                spark_association_database = spark.read.parquet(database_location+\"associationByOverallDirect/\")\n",
    "            else:\n",
    "                datasource_filter = True\n",
    "                spark_association_database_pre = spark.read.parquet(database_location+\"associationByDatasourceDirect/\")\n",
    "        else: raise Exception('Unknown error loading database')#should raise but put it there in case    \n",
    "        # associationByDatatypeDirect = spark.read.parquet(database_location+\"associationByDatatypeDirect/\") #we are not including datatypes yet\n",
    "        # associationByDatatypeIndirect = spark.read.parquet(database_location+\"associationByDatatypeIndirect/\")\n",
    "    except ValueError:\n",
    "        print('error loading database, make sure the database location contains the required files')\n",
    "    \n",
    "    #check if disease ID exists\n",
    "    if not bool(diseases.filter(col(\"id\")==disease_ID).collect()):\n",
    "        raise ValueError('Disease ID does not exist')\n",
    "    \n",
    "    #create disease list\n",
    "    disease_list = [disease_ID] #disease list starts with the disease ID\n",
    "    if include_descendants: #if descendants are to be included\n",
    "        print('including descendants')\n",
    "        descendant_list = diseases.filter(col(\"id\")==disease_ID).select('descendants',).first()[0] #list of descendants for specific disease id\n",
    "        print('found', len(descendant_list), 'descendants')\n",
    "        disease_list = disease_list + descendant_list #concatenate two lists\n",
    "        print('list of disease_IDs:',disease_list)\n",
    "    \n",
    "    #Get list of genes for diseases\n",
    "    if datasource_filter is True:\n",
    "        spark_association_database = spark_association_database_pre.filter(col(\"datasourceId\").isin(datasource))\n",
    "        print('filtering by datasource list:',datasource)\n",
    "        spark_association_database.persist() #persist for quicker loading\n",
    "    else:\n",
    "        spark_association_database.persist()\n",
    "    \n",
    "    #getting list of genes\n",
    "    total_genes = (spark_association_database\n",
    "                           .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "                           .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "                           .dropDuplicates(['targetId'])\n",
    "                           .count()\n",
    "                          )\n",
    "    print(total_genes, 'genes found')\n",
    "    if gene_cutoff == 'all' or gene_cutoff == 'All': #keep all genes\n",
    "        gene_list = list((spark_association_database\n",
    "                          .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "                          .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "                          .dropDuplicates(['targetId']) #drop duplicates\n",
    "                          .orderBy('score', ascending = False) #?order again\n",
    "                          .select('targetId')\n",
    "                          .toPandas()\n",
    "                         )['targetId'])\n",
    "        print('keeping all genes')\n",
    "    elif isinstance(gene_cutoff, float): #score cutoff filter\n",
    "        if (gene_cutoff>=0)&(gene_cutoff<=1): #make sure the score cutoff is between 0 and 1\n",
    "            gene_list = list((spark_association_database\n",
    "                              .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "                              .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "                              .dropDuplicates(['targetId']) #drop duplicates\n",
    "                              .filter(col(\"score\")>=gene_cutoff)\n",
    "                              .orderBy('score', ascending = False) #?order again\n",
    "                              .select('targetId')\n",
    "                              .toPandas()      \n",
    "                             )['targetId'])\n",
    "            print('filtering scores larger than', gene_cutoff)\n",
    "            print('keeping', len(gene_list), 'genes')\n",
    "        else: raise Exception(\"gene_cutoff needs to be 'All', a float between 0 and 1 or an int\")\n",
    "    elif isinstance(gene_cutoff, int): #top n genes filter\n",
    "        gene_list = list((spark_association_database\n",
    "                          .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "                          .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "                          .dropDuplicates(['targetId']) #drop duplicates\n",
    "                          .orderBy('score', ascending = False) #?order again\n",
    "                          .limit(gene_cutoff)\n",
    "                          .select('targetId')\n",
    "                          .toPandas()      \n",
    "                       )['targetId'])\n",
    "        print('filtering top', gene_cutoff, 'genes')\n",
    "        print('keeping', len(gene_list), 'genes')\n",
    "    else: raise Exception('Unknown error filtering genes') #idk what would cause this to raise but I put it there in case\n",
    "    \n",
    "    diseases.unpersist() #unload data from cache\n",
    "    spark_association_database.unpersist()\n",
    "\n",
    "    #map orthologues\n",
    "    if ortholog_map is not None:\n",
    "        print('mapping using specified orthologue file')\n",
    "        ortholog_map_df = pd.read_csv(ortholog_map)\n",
    "        ortholog_map_df = ortholog_map_df.dropna()\n",
    "        input_gene_count = len(gene_list)\n",
    "        if return_gene_names is True:\n",
    "            ortholog_map_dict = dict(zip(ortholog_map_df.iloc[:,0], ortholog_map_df.iloc[:,3]))\n",
    "        else:\n",
    "            ortholog_map_dict = dict(zip(ortholog_map_df.iloc[:,0], ortholog_map_df.iloc[:,2]))\n",
    "        gene_list = [x for x in map(ortholog_map_dict.get, gene_list) if x is not None]\n",
    "        print(len(gene_list), 'focal species genes mapped to', input_gene_count, 'human genes.')\n",
    "        \n",
    "    return(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aa5dfcf3-3106-43a1-992d-7eaa42a23562",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#gwas query function\n",
    "#Supply a disease code, check the database for associated genes. Cutoff either by score or by top n genes. Can use overall score, limit to results from specific datasource, \n",
    "#use database score. List of genes is human, needs to be moved to mouse ortholog.\n",
    "\n",
    "###code step by step\n",
    "\n",
    "#done - check if OT database is present, if not error\n",
    "#done - Check if disease code exists in database, if not report error\n",
    "#done - check if datasource exists (if not using overall score), if not report error\n",
    "    #either direct or indirect\n",
    "#done - Include disease children yes/no ###be sure to include original term as well\n",
    "    #if yes, report the codes of the children\n",
    "#done - Sort by score of datasource of choice\n",
    "#done - drop duplicates in correct order\n",
    "#done - report total unique genes found\n",
    "#compare with mouse, find orthologs\n",
    "    #report number of orthologs found\n",
    "    #if we want the gene name, change ensemblid to gene name\n",
    "    #########################\n",
    "    #this part needs some work, need to properly drop NA only in the ID column and replace missing gene names with ID. Now we might toss genes that do map.\n",
    "    #need to write a function to retrieve ensembl orthologs by itself\n",
    "    #need to write a case for human genes to skip orthologs but still need to map them to gene names if wanted\n",
    "\n",
    "#Filter by selected method, (all, score cut-off (between 0-1), top n genes)\n",
    "#return list of genes (either by ensembl id or gene name)\n",
    "\n",
    "############\n",
    "#once finished I need to rewrite the code and take out some of the subfunctions for readability and easy understanding\n",
    "#should create a class with all the required subfunctions\n",
    "\n",
    "\n",
    "#### import libraries ####\n",
    "import os\n",
    "import pyspark\n",
    "from pyspark.sql import SparkSession\n",
    "from pyspark.sql.functions import col, explode\n",
    "from typing import Union\n",
    "\n",
    "def OT_genes_by_disease(\n",
    "    disease_ID: str = 'MONDO_0016064', #disease id, set default to cleft palate\n",
    "    gene_cutoff: Union[int, float, str] = 'All', #All for all genes, cutoff score between 0-1, or top n genes cutoff\n",
    "    datasource: list = ['overall'], #somehow check if datasource exists, default uses overall. list of datasources\n",
    "    evidence_type: str = 'indirect', #whether to use direct or indirect evidence\n",
    "    database_location: str = 'GWAS/opentargets/platform/', #location of the database\n",
    "    include_descendants: bool = True, #whether or not to include disease descendants\n",
    "    ortholog_map: str = 'GWAS/ensembl_human_orthologues/Human_to_Mouse.txt', #Location of csv file mapping human genes to focal species. Expects columns in the following order, Human gene stable ID, Human gene name, Focal species stable ID, focal species gene name.\n",
    "    return_gene_names: bool = True, #whether to return gene names or ensemblid\n",
    "\n",
    "    \n",
    "):   \n",
    "    #some checks on variables\n",
    "    datasource_list = ['overall','europepmc', 'crispr', 'uniprot_variants', 'genomics_england', 'gene_burden', 'chembl', 'slapenrich', 'crispr_screen', \n",
    "                       'expression_atlas','intogen', 'reactome', 'orphanet', 'sysbio', 'cancer_biomarkers', 'cancer_gene_census', 'progeny', 'clingen',\n",
    "                       'eva', 'gene2phenotype', 'uniprot_literature','eva_somatic', 'ot_genetics_portal', 'impc']\n",
    "    if datasource not in datasource_list:\n",
    "        if not set(datasource).issubset(datasource_list):\n",
    "            raise ValueError(\"Datasource does not exist. Make sure datasource is 'overall' or a valid datasource.\")\n",
    "    \n",
    "    evidence_type_list = ['direct', 'indirect']\n",
    "    if evidence_type not in evidence_type_list:\n",
    "        raise ValueError(\"evidence_type needs to be either 'direct' or 'indirect'.\")\n",
    "    \n",
    "    #test if database directory exists\n",
    "    if not os.path.isdir(database_location):\n",
    "        raise ValueError('Database directory does not exist')\n",
    "    \n",
    "    #start spark session\n",
    "    spark = SparkSession.builder \\\n",
    "            .getOrCreate()\n",
    "    \n",
    "    #load database\n",
    "    try:\n",
    "        diseases = spark.read.parquet(database_location+\"diseases/\") #load diseases reference\n",
    "        diseases.persist() #persist diseases for quicker loading\n",
    "        #now we want to load the correct database\n",
    "        datasource_filter = False\n",
    "        if evidence_type=='indirect': #if we want indirect evidence\n",
    "            if datasource==['overall']: #if we want to look at overall scores\n",
    "                spark_association_database = spark.read.parquet(database_location+\"associationByOverallIndirect/\")\n",
    "            else: #if we want to look at other datasource scores\n",
    "                datasource_filter = True                \n",
    "                spark_association_database_pre = spark.read.parquet(database_location+\"associationByDatasourceIndirect/\") #name it pre so we can filter afterwards and use the same code\n",
    "        elif evidence_type=='direct':#if we want direct evidence\n",
    "            if datasource==['overall']: \n",
    "                spark_association_database = spark.read.parquet(database_location+\"associationByOverallDirect/\")\n",
    "            else:\n",
    "                datasource_filter = True\n",
    "                spark_association_database_pre = spark.read.parquet(database_location+\"associationByDatasourceDirect/\")\n",
    "        else: raise Exception('Unknown error loading database')#should raise but put it there in case    \n",
    "        # associationByDatatypeDirect = spark.read.parquet(database_location+\"associationByDatatypeDirect/\") #we are not including datatypes yet\n",
    "        # associationByDatatypeIndirect = spark.read.parquet(database_location+\"associationByDatatypeIndirect/\")\n",
    "    except ValueError:\n",
    "        print('error loading database, make sure the database location contains the required files')\n",
    "    \n",
    "    #check if disease ID exists\n",
    "    if not bool(diseases.filter(col(\"id\")==disease_ID).collect()):\n",
    "        raise ValueError('Disease ID does not exist')\n",
    "    \n",
    "    #create disease list\n",
    "    disease_list = [disease_ID] #disease list starts with the disease ID\n",
    "    if include_descendants: #if descendants are to be included\n",
    "        print('including descendants')\n",
    "        descendant_list = diseases.filter(col(\"id\")==disease_ID).select('descendants',).first()[0] #list of descendants for specific disease id\n",
    "        print('found', len(descendant_list), 'descendants')\n",
    "        disease_list = disease_list + descendant_list #concatenate two lists\n",
    "        print('list of disease_IDs:',disease_list)\n",
    "    \n",
    "    #Get list of genes for diseases\n",
    "    if datasource_filter is True:\n",
    "        spark_association_database = spark_association_database_pre.filter(col(\"datasourceId\").isin(datasource))\n",
    "        print('filtering by datasource list:',datasource)\n",
    "        spark_association_database.persist() #persist for quicker loading\n",
    "    else:\n",
    "        spark_association_database.persist()\n",
    "    \n",
    "    #getting list of genes\n",
    "    total_genes = (spark_association_database\n",
    "                           .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "                           .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "                           .dropDuplicates(['targetId'])\n",
    "                           .count()\n",
    "                          )\n",
    "    print(total_genes, 'genes found')\n",
    "    if gene_cutoff == 'all' or gene_cutoff == 'All': #keep all genes\n",
    "        gene_list_df = (spark_association_database\n",
    "                          .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "                          .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "                          .dropDuplicates(['targetId']) #drop duplicates\n",
    "                          .orderBy('score', ascending = False) #?order again\n",
    "                          .select('targetId','score')\n",
    "                          .toPandas()\n",
    "                         )\n",
    "        print('keeping all genes')\n",
    "    elif isinstance(gene_cutoff, float): #score cutoff filter\n",
    "        if (gene_cutoff>=0)&(gene_cutoff<=1): #make sure the score cutoff is between 0 and 1\n",
    "            gene_list_df = (spark_association_database\n",
    "                              .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "                              .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "                              .dropDuplicates(['targetId']) #drop duplicates\n",
    "                              .filter(col(\"score\")>=gene_cutoff)\n",
    "                              .orderBy('score', ascending = False) #?order again\n",
    "                              .select('targetId','score')\n",
    "                              .toPandas()      \n",
    "                             )\n",
    "            print('filtering scores larger than', gene_cutoff)\n",
    "            print('keeping', len(gene_list_df), 'genes')\n",
    "        else: raise Exception(\"gene_cutoff needs to be 'All', a float between 0 and 1 or an int\")\n",
    "    elif isinstance(gene_cutoff, int): #top n genes filter\n",
    "        gene_list_df = (spark_association_database\n",
    "                          .filter(col(\"diseaseId\").isin(disease_list))  #filter by disease ID\n",
    "                          .orderBy('score', ascending = False) #reorder by score before removing duplicates\n",
    "                          .dropDuplicates(['targetId']) #drop duplicates\n",
    "                          .orderBy('score', ascending = False) #?order again\n",
    "                          .limit(gene_cutoff)\n",
    "                          .select('targetId','score')\n",
    "                          .toPandas()      \n",
    "                       )\n",
    "        print('filtering top', gene_cutoff, 'genes')\n",
    "        print('keeping', len(gene_list_df), 'genes')\n",
    "    else: raise Exception('Unknown error filtering genes') #idk what would cause this to raise but I put it there in case\n",
    "    \n",
    "    diseases.unpersist() #unload data from cache\n",
    "    spark_association_database.unpersist()\n",
    "\n",
    "    #map orthologues\n",
    "    if ortholog_map is not None:\n",
    "        print('mapping using specified orthologue file')\n",
    "        ortholog_map_df = pd.read_csv(ortholog_map)\n",
    "        ortholog_map_df = ortholog_map_df.dropna()\n",
    "        gene_list = list(gene_list_df['targetId'])\n",
    "        input_gene_count = len(gene_list)\n",
    "        if return_gene_names is True:\n",
    "            ortholog_map_dict = dict(zip(ortholog_map_df.iloc[:,0], ortholog_map_df.iloc[:,3]))\n",
    "        else:\n",
    "            ortholog_map_dict = dict(zip(ortholog_map_df.iloc[:,0], ortholog_map_df.iloc[:,2]))\n",
    "        #gene_list = [x for x in map(ortholog_map_dict.get, gene_list) if x is not None]\n",
    "        gene_list_df['targetId'] = gene_list_df['targetId'].map(ortholog_map_dict)\n",
    "        gene_list_df.dropna(inplace=True)\n",
    "        print(len(gene_list_df), 'focal species genes mapped to', input_gene_count, 'human genes.')\n",
    "        gene_list_df = gene_list_df.to_dict()\n",
    "        #gene_list_df = gene_list_df[gene_list_df['targetId'].isin(gene_list)]\n",
    "    return(gene_list_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "1b6f8bbb-8b03-492a-82f6-852269ea9a29",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "including descendants\n",
      "found 63 descendants\n",
      "list of disease_IDs: ['MONDO_0023369', 'MONDO_0021580', 'MONDO_0004328', 'EFO_0007356', 'MONDO_0024611', 'MONDO_0002580', 'MONDO_0029145', 'MONDO_0002186', 'EFO_1000209', 'MONDO_0015411', 'MONDO_0012142', 'EFO_1001818', 'MONDO_0015416', 'MONDO_0015417', 'EFO_1000365', 'MONDO_0016064', 'MONDO_0015496', 'Orphanet_3253', 'MONDO_0010560', 'MONDO_0001122', 'EFO_0007264', 'MONDO_0000110', 'MONDO_0001123', 'MONDO_0004312', 'MONDO_0008637', 'EFO_1001077', 'MONDO_0002579', 'MONDO_0000358', 'MONDO_0004757', 'MONDO_0004394', 'MONDO_0017627', 'MONDO_0015480', 'EFO_1000731', 'MONDO_0015481', 'MONDO_0013300', 'EFO_0007333', 'MONDO_0008866', 'MONDO_0004943', 'MONDO_0004384', 'MONDO_0007336', 'MONDO_0021484', 'MONDO_0010906', 'MONDO_0012165', 'MONDO_0014388', 'MONDO_0016043', 'MONDO_0016044', 'EFO_0007408', 'EFO_1001035', 'EFO_0700035', 'MONDO_0013378', 'MONDO_0004810', 'MONDO_0004457', 'MONDO_0007927', 'EFO_0007489', 'MONDO_0001624', 'MONDO_0021412', 'EFO_0007360', 'EFO_0007361', 'MONDO_0009151', 'MONDO_0015824', 'MONDO_0001748', 'MONDO_0010850', 'MONDO_0015420', 'MONDO_0014772']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1299 genes found\n",
      "filtering scores larger than 0.2\n",
      "keeping 59 genes\n",
      "mapping using specified orthologue file\n",
      "58 focal species genes mapped to 59 human genes.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "58"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trait_gene_dict = OT_genes_by_disease(disease_ID = 'MONDO_0023369',\n",
    "                                             datasource = ['overall'],\n",
    "                                             gene_cutoff = 0.2,)\n",
    "list(trait_gene_dict['targetId'].values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "3e7cf898-5e1a-4569-9b75-c775ccfbad6a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "including descendants\n",
      "found 2 descendants\n",
      "list of disease_IDs: ['EFO_0003959', 'HP_0000204', 'HP_0000161']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "455 genes found\n",
      "keeping all genes\n",
      "mapping using specified orthologue file\n",
      "433 focal species genes mapped to 455 human genes.\n",
      "including descendants\n",
      "found 0 descendants\n",
      "list of disease_IDs: ['HP_0000175']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "759 genes found\n",
      "keeping all genes\n",
      "mapping using specified orthologue file\n",
      "728 focal species genes mapped to 759 human genes.\n",
      "including descendants\n",
      "found 0 descendants\n",
      "list of disease_IDs: ['EFO_0007845']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "67 genes found\n",
      "keeping all genes\n",
      "mapping using specified orthologue file\n",
      "66 focal species genes mapped to 67 human genes.\n",
      "including descendants\n",
      "found 0 descendants\n",
      "list of disease_IDs: ['EFO_0007842']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "80 genes found\n",
      "keeping all genes\n",
      "mapping using specified orthologue file\n",
      "78 focal species genes mapped to 80 human genes.\n",
      "including descendants\n",
      "found 0 descendants\n",
      "list of disease_IDs: ['EFO_0007843']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "158 genes found\n",
      "keeping all genes\n",
      "mapping using specified orthologue file\n",
      "152 focal species genes mapped to 158 human genes.\n",
      "including descendants\n",
      "found 5 descendants\n",
      "list of disease_IDs: ['EFO_0005410', 'MONDO_0007129', 'MONDO_0014749', 'MONDO_0007881', 'MONDO_0010741', 'MONDO_0011265']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "339 genes found\n",
      "keeping all genes\n",
      "mapping using specified orthologue file\n",
      "321 focal species genes mapped to 339 human genes.\n",
      "including descendants\n",
      "found 14 descendants\n",
      "list of disease_IDs: ['EFO_0007841', 'EFO_0007842', 'EFO_0007843', 'EFO_0009360', 'EFO_0010949', 'EFO_0010948', 'EFO_0009892', 'EFO_0007844', 'EFO_0007955', 'EFO_0007856', 'EFO_0007845', 'EFO_0007855', 'EFO_0007858', 'EFO_0007857', 'EFO_0007847']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "542 genes found\n",
      "keeping all genes\n",
      "mapping using specified orthologue file\n",
      "524 focal species genes mapped to 542 human genes.\n"
     ]
    }
   ],
   "source": [
    "###running on dict of traits, produce dict of genes\n",
    "#save as json\n",
    "trait_list = {'cleft_lip' :'EFO_0003959',\n",
    "              'cleft_palate' : 'HP_0000175',\n",
    "              'lip_morphology_measurement' : 'EFO_0007845',\n",
    "              'chin_morphology_measurement' : 'EFO_0007842',\n",
    "              'nose_morphology_measurement' : 'EFO_0007843',\n",
    "              'tooth_agenesis' : 'EFO_0005410',\n",
    "              'facial_morphology_measurement':'EFO_0007841',}\n",
    "\n",
    "trait_gene_dict = {}\n",
    "for i in trait_list:\n",
    "    trait_gene_dict[i] = OT_genes_by_disease(disease_ID = trait_list[i],\n",
    "                                             datasource = ['overall'],\n",
    "                                             gene_cutoff = 'all',)\n",
    "    # if len(trait_gene_dict[i]['targetId']) < 50:\n",
    "    #     trait_gene_dict[i] = OT_genes_by_disease(disease_ID = trait_list[i],\n",
    "    #                                              datasource = ['overall'],\n",
    "    #                                              gene_cutoff = 50,)\n",
    "\n",
    "import json\n",
    "with open('GWAS/trait_json/test_traits_4.json', 'w',encoding='utf-8') as f: \n",
    "    json.dump(trait_gene_dict, f, ensure_ascii=False, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "ff471a05-9a91-4211-9a73-a3299d392944",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('GWAS/trait_json/test_traits_2.json', 'w',encoding='utf-8') as f: \n",
    "    json.dump(trait_gene_dict, f, ensure_ascii=False, indent=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33d71611-89c9-43c7-b23a-5689681f75c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list_df = evidence_ot_genetics_portal.filter(col(\"studyId\").isin(list(studyID_dict.keys()))).filter(col(\"score\")>min_score).select(['targetId','variantRsId','studyId','score','diseaseFromSourceMappedId']).toPandas()\n",
    "gene_list_df['area'] = gene_list_df['studyId'].map(area_dict)\n",
    "gene_list_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "8c1e2bb8-9015-4264-a3c2-05fee54141d6",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "90 focal species genes mapped to 94 human genes.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetId</th>\n",
       "      <th>variantRsId</th>\n",
       "      <th>studyId</th>\n",
       "      <th>score</th>\n",
       "      <th>diseaseFromSourceMappedId</th>\n",
       "      <th>area</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Tmem74</td>\n",
       "      <td>rs10089785</td>\n",
       "      <td>GCST90007253</td>\n",
       "      <td>0.690749</td>\n",
       "      <td>EFO_0007843</td>\n",
       "      <td>Nose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Nav3</td>\n",
       "      <td>rs2695152</td>\n",
       "      <td>GCST90007215</td>\n",
       "      <td>0.656231</td>\n",
       "      <td>EFO_0007845</td>\n",
       "      <td>Upper_lip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gm525</td>\n",
       "      <td>rs227727</td>\n",
       "      <td>GCST90007218</td>\n",
       "      <td>0.346529</td>\n",
       "      <td>EFO_0007845</td>\n",
       "      <td>Upper_lip</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pkdcc</td>\n",
       "      <td>rs10189338</td>\n",
       "      <td>GCST90007191</td>\n",
       "      <td>0.714584</td>\n",
       "      <td>EFO_0007843</td>\n",
       "      <td>Nose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Hgf</td>\n",
       "      <td>rs7806852</td>\n",
       "      <td>GCST90007248</td>\n",
       "      <td>0.700607</td>\n",
       "      <td>EFO_0007843</td>\n",
       "      <td>Nose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>Has2</td>\n",
       "      <td>rs2581548</td>\n",
       "      <td>GCST90007245</td>\n",
       "      <td>0.483594</td>\n",
       "      <td>EFO_0007841</td>\n",
       "      <td>Frontonasal</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>Runx2</td>\n",
       "      <td>rs73457129</td>\n",
       "      <td>GCST90007264</td>\n",
       "      <td>0.788357</td>\n",
       "      <td>EFO_0007843</td>\n",
       "      <td>Nose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>Sox9</td>\n",
       "      <td>rs9302943</td>\n",
       "      <td>GCST90007224</td>\n",
       "      <td>0.744489</td>\n",
       "      <td>EFO_0007843</td>\n",
       "      <td>Nose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>Six2</td>\n",
       "      <td>rs35395759</td>\n",
       "      <td>GCST90007224</td>\n",
       "      <td>0.627206</td>\n",
       "      <td>EFO_0007843</td>\n",
       "      <td>Nose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93</th>\n",
       "      <td>Eefsec</td>\n",
       "      <td>rs7373685</td>\n",
       "      <td>GCST90007261</td>\n",
       "      <td>0.488346</td>\n",
       "      <td>EFO_0007845</td>\n",
       "      <td>Upper_lip</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   targetId variantRsId       studyId     score diseaseFromSourceMappedId  \\\n",
       "0    Tmem74  rs10089785  GCST90007253  0.690749               EFO_0007843   \n",
       "1      Nav3   rs2695152  GCST90007215  0.656231               EFO_0007845   \n",
       "2     Gm525    rs227727  GCST90007218  0.346529               EFO_0007845   \n",
       "3     Pkdcc  rs10189338  GCST90007191  0.714584               EFO_0007843   \n",
       "4       Hgf   rs7806852  GCST90007248  0.700607               EFO_0007843   \n",
       "..      ...         ...           ...       ...                       ...   \n",
       "89     Has2   rs2581548  GCST90007245  0.483594               EFO_0007841   \n",
       "90    Runx2  rs73457129  GCST90007264  0.788357               EFO_0007843   \n",
       "91     Sox9   rs9302943  GCST90007224  0.744489               EFO_0007843   \n",
       "92     Six2  rs35395759  GCST90007224  0.627206               EFO_0007843   \n",
       "93   Eefsec   rs7373685  GCST90007261  0.488346               EFO_0007845   \n",
       "\n",
       "           area  \n",
       "0          Nose  \n",
       "1     Upper_lip  \n",
       "2     Upper_lip  \n",
       "3          Nose  \n",
       "4          Nose  \n",
       "..          ...  \n",
       "89  Frontonasal  \n",
       "90         Nose  \n",
       "91         Nose  \n",
       "92         Nose  \n",
       "93    Upper_lip  \n",
       "\n",
       "[90 rows x 6 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ortholog_map = 'GWAS/ensembl_human_orthologues/Human_to_Mouse.txt'\n",
    "\n",
    "ortholog_map_df = pd.read_csv(ortholog_map)\n",
    "ortholog_map_df = ortholog_map_df.dropna()\n",
    "gene_list = list(gene_list_df['targetId'])\n",
    "input_gene_count = len(gene_list)\n",
    "ortholog_map_dict = dict(zip(ortholog_map_df.iloc[:,0], ortholog_map_df.iloc[:,3]))\n",
    "gene_list_df['targetId'] = gene_list_df['targetId'].map(ortholog_map_dict)\n",
    "\n",
    "#gene_list = [x for x in map(ortholog_map_dict.get, gene_list) if x is not None]\n",
    "gene_list_df = gene_list_df[gene_list_df['targetId'].notna()]\n",
    "print(len(gene_list_df), 'focal species genes mapped to', input_gene_count, 'human genes.')\n",
    "gene_list_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "2902b95a-e8a9-439d-b8db-9e95ce1b1e0e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gene_list_df.sort_values('studyId').to_csv('White_JD_genes_df.csv',index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bdaa2e1b-1a1a-4085-90e7-4228856cf24c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "including descendants\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "found 14 descendants\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                "
     ]
    }
   ],
   "source": [
    "disease_ID = 'EFO_0007841' #MONDO_0023369\n",
    "min_score = 0.1\n",
    "\n",
    "disease_list = [disease_ID] #disease list starts with the disease ID\n",
    "print('including descendants')\n",
    "descendant_list = diseases.filter(col(\"id\")==disease_ID).select('descendants',).first()[0] #list of descendants for specific disease id\n",
    "print('found', len(descendant_list), 'descendants')\n",
    "disease_list = disease_list + descendant_list #concatenate two lists\n",
    "gene_list_df = evidence_ot_genetics_portal.filter(col(\"diseaseFromSourceMappedId\").isin(disease_list)).filter(col(\"score\")>min_score).select(['targetId','variantRsId','studyId','score','diseaseFromSourceMappedId']).toPandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1ac51e2b-5e3c-4fcb-9964-ce69551ce100",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Mouse gene stable ID</th>\n",
       "      <th>Mouse gene name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77642</th>\n",
       "      <td>ENSG00000232679</td>\n",
       "      <td>LINC01705</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77643</th>\n",
       "      <td>ENSG00000200033</td>\n",
       "      <td>RNU6-403P</td>\n",
       "      <td>ENSMUSG00000088001</td>\n",
       "      <td>Gm22883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77644</th>\n",
       "      <td>ENSG00000228437</td>\n",
       "      <td>LINC02474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77645</th>\n",
       "      <td>ENSG00000229463</td>\n",
       "      <td>LYST-AS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77646</th>\n",
       "      <td>ENSG00000229291</td>\n",
       "      <td>LINC02768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77647 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Gene stable ID  Gene name Mouse gene stable ID Mouse gene name\n",
       "0      ENSG00000210049      MT-TF                  NaN             NaN\n",
       "1      ENSG00000211459    MT-RNR1                  NaN             NaN\n",
       "2      ENSG00000210077      MT-TV                  NaN             NaN\n",
       "3      ENSG00000210082    MT-RNR2                  NaN             NaN\n",
       "4      ENSG00000209082     MT-TL1                  NaN             NaN\n",
       "...                ...        ...                  ...             ...\n",
       "77642  ENSG00000232679  LINC01705                  NaN             NaN\n",
       "77643  ENSG00000200033  RNU6-403P   ENSMUSG00000088001         Gm22883\n",
       "77644  ENSG00000228437  LINC02474                  NaN             NaN\n",
       "77645  ENSG00000229463   LYST-AS1                  NaN             NaN\n",
       "77646  ENSG00000229291  LINC02768                  NaN             NaN\n",
       "\n",
       "[77647 rows x 4 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ortholog_map = 'GWAS/ensembl_human_orthologues/Human_to_Mouse.txt'\n",
    "\n",
    "ortholog_map_df = pd.read_csv(ortholog_map)\n",
    "ortholog_map_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6b0e6d1a-17f2-4848-95c1-dbc5594c2a56",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "including descendants\n",
      "found 14 descendants\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 31:================================>                      (31 + 22) / 53]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "356 focal species genes mapped to 372 human genes.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                "
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetId</th>\n",
       "      <th>variantRsId</th>\n",
       "      <th>studyId</th>\n",
       "      <th>score</th>\n",
       "      <th>diseaseFromSourceMappedId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Cped1</td>\n",
       "      <td>rs884373</td>\n",
       "      <td>GCST90007206</td>\n",
       "      <td>0.811751</td>\n",
       "      <td>EFO_0010948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cped1</td>\n",
       "      <td>rs6950680</td>\n",
       "      <td>GCST90026584</td>\n",
       "      <td>0.788809</td>\n",
       "      <td>EFO_0010948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Bmp2</td>\n",
       "      <td>rs3920540</td>\n",
       "      <td>GCST006517</td>\n",
       "      <td>0.439235</td>\n",
       "      <td>EFO_0007843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Tmem74</td>\n",
       "      <td>rs10089785</td>\n",
       "      <td>GCST90007253</td>\n",
       "      <td>0.690749</td>\n",
       "      <td>EFO_0007843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Alx4</td>\n",
       "      <td>rs7121535</td>\n",
       "      <td>GCST90007181</td>\n",
       "      <td>0.867678</td>\n",
       "      <td>EFO_0007841</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367</th>\n",
       "      <td>Six2</td>\n",
       "      <td>rs921120</td>\n",
       "      <td>GCST90026563</td>\n",
       "      <td>0.761771</td>\n",
       "      <td>EFO_0007843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368</th>\n",
       "      <td>Six2</td>\n",
       "      <td>rs35395759</td>\n",
       "      <td>GCST90007224</td>\n",
       "      <td>0.627206</td>\n",
       "      <td>EFO_0007843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369</th>\n",
       "      <td>Eefsec</td>\n",
       "      <td>rs7373685</td>\n",
       "      <td>GCST90007261</td>\n",
       "      <td>0.488346</td>\n",
       "      <td>EFO_0007845</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370</th>\n",
       "      <td>Fgf9</td>\n",
       "      <td>rs73168105</td>\n",
       "      <td>GCST003989</td>\n",
       "      <td>0.455364</td>\n",
       "      <td>EFO_0007842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>Hlx</td>\n",
       "      <td>rs12039502</td>\n",
       "      <td>GCST90007193</td>\n",
       "      <td>0.727691</td>\n",
       "      <td>EFO_0010948</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>356 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    targetId variantRsId       studyId     score diseaseFromSourceMappedId\n",
       "0      Cped1    rs884373  GCST90007206  0.811751               EFO_0010948\n",
       "1      Cped1   rs6950680  GCST90026584  0.788809               EFO_0010948\n",
       "2       Bmp2   rs3920540    GCST006517  0.439235               EFO_0007843\n",
       "3     Tmem74  rs10089785  GCST90007253  0.690749               EFO_0007843\n",
       "4       Alx4   rs7121535  GCST90007181  0.867678               EFO_0007841\n",
       "..       ...         ...           ...       ...                       ...\n",
       "367     Six2    rs921120  GCST90026563  0.761771               EFO_0007843\n",
       "368     Six2  rs35395759  GCST90007224  0.627206               EFO_0007843\n",
       "369   Eefsec   rs7373685  GCST90007261  0.488346               EFO_0007845\n",
       "370     Fgf9  rs73168105    GCST003989  0.455364               EFO_0007842\n",
       "371      Hlx  rs12039502  GCST90007193  0.727691               EFO_0010948\n",
       "\n",
       "[356 rows x 5 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_ID = 'EFO_0007841' #MONDO_0023369\n",
    "min_score = 0.4\n",
    "\n",
    "disease_list = [disease_ID] #disease list starts with the disease ID\n",
    "print('including descendants')\n",
    "descendant_list = diseases.filter(col(\"id\")==disease_ID).select('descendants',).first()[0] #list of descendants for specific disease id\n",
    "print('found', len(descendant_list), 'descendants')\n",
    "disease_list = disease_list + descendant_list #concatenate two lists\n",
    "gene_list_df = evidence_ot_genetics_portal.filter(col(\"diseaseFromSourceMappedId\").isin(disease_list)).filter(col(\"score\")>min_score).select(['targetId','variantRsId','studyId','score','diseaseFromSourceMappedId']).toPandas()\n",
    "ortholog_map = 'GWAS/ensembl_human_orthologues/Human_to_Mouse.txt'\n",
    "\n",
    "ortholog_map_df = pd.read_csv(ortholog_map)\n",
    "ortholog_map_df = ortholog_map_df.dropna()\n",
    "gene_list = list(gene_list_df['targetId'])\n",
    "input_gene_count = len(gene_list)\n",
    "ortholog_map_dict = dict(zip(ortholog_map_df.iloc[:,0], ortholog_map_df.iloc[:,3]))\n",
    "gene_list_df['targetId'] = gene_list_df['targetId'].map(ortholog_map_dict)\n",
    "\n",
    "#gene_list = [x for x in map(ortholog_map_dict.get, gene_list) if x is not None]\n",
    "gene_list_df = gene_list_df[gene_list_df['targetId'].notna()]\n",
    "print(len(gene_list_df), 'focal species genes mapped to', input_gene_count, 'human genes.')\n",
    "gene_list_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "faae0205-a023-44e6-9af7-19f4eb321221",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease_dict = {\n",
    "    'EFO_0007841': 'facial morphology measurement(GWAS)',\n",
    "    'HP_0000202': 'orofacial cleft(GWAS)',\n",
    "    'HP_0000366': 'abnormality of the nose',\n",
    "    'MONDO_0000358': 'orofacial cleft',\n",
    "    'HP_0001999': 'abnormal of facial shape',\n",
    "    'HP_0000153': 'abnormality of the mouth',\n",
    "    'HP_0000164': 'abnormality of dentition',\n",
    "    'MONDO_0016643': 'frontonasal dysplasia',\n",
    "    'EFO_1000720': 'keratosis',\n",
    "    'EFO_0005917': 'generalised epilepsy',\n",
    "    'MONDO_0003939': 'muscle tissue disorder',\n",
    "    'EFO_0000538':'hypertrophic cardiomyopathy',\n",
    "    'EFO_0000318':'cardiomyopathy',\n",
    "    'MONDO_0020120':'skeletal muscle disorder',\n",
    "    'MONDO_0019952':'congenital myopathy',\n",
    "    'EFO_0000540':'immune system disease',\n",
    "    'EFO_0000407':'dilated cardiomyopathy(GWAS)',\n",
    "    'EFO_0006941':'grip strength measurement(GWAS)',\n",
    "    'EFO_0004264':'vascular disease',\n",
    "    'EFO_0004695':'intraocular pressure',\n",
    "    'EFO_0008003':'heart rate variability measurement(GWAS)',\n",
    "    'EFO_0006336':'diastolic blood pressure',\n",
    "    'MONDO_0005180':'parkinsons disease',\n",
    "    'MONDO_0020127':'hereditary peripheral neuropathy',\n",
    "    'MONDO_0015626':'Charcot-Marie-Tooth disease',\n",
    "    'MONDO_0019548':'autosomal dominant intermediate Charcot-Marie-Tooth disease',\n",
    "    'MONDO_0024237':'inherited neurodegenerative disorder',\n",
    "    'MONDO_0100172':'intellectual disability, autosomal dominant',\n",
    "    'EFO_0005772':'neurodegenerative disease',\n",
    "    'EFO_0010642':'neurodevelopmental disorder',\n",
    "    'EFO_0004875':'mathematical ability',\n",
    "    'MONDO_0023369':'disorder of facial skeleton',\n",
    "    'HP_0011821':'abnormal facial skeleton morphology',\n",
    "    'HP_0000271':'abnormality of the face',\n",
    "    'MONDO_0015411': 'facial cleft',\n",
    "    'HP_0001742':'nasal congestion',\n",
    "}\n",
    "\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "6f4918c0-e828-4eb4-ad42-a029643b5c0a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "including descendants\n",
      "found 0 descendants\n",
      "20 focal species genes mapped to 20 human genes.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetId</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adrb1</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.540921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Slc6a2</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.603508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Trpa1</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.565428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Nr3c1</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.588986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Adra1a</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.606803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Avpr2</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.564504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Chrm3</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.565059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Trpm8</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.564504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Sigmar1</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.606034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Adra2a</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.600616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Avpr1a</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.564504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Adrb2</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.540921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Adra1b</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.606796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Adra1d</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.607350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Grin2a</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.606034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Adra2c</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.600616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Adrb3</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.540921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Hrh1</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.606959</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Avpr1b</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.564504</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Adra2b</td>\n",
       "      <td>HP_0001742</td>\n",
       "      <td>0.600616</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   targetId   diseaseId     score\n",
       "0     Adrb1  HP_0001742  0.540921\n",
       "1    Slc6a2  HP_0001742  0.603508\n",
       "2     Trpa1  HP_0001742  0.565428\n",
       "3     Nr3c1  HP_0001742  0.588986\n",
       "4    Adra1a  HP_0001742  0.606803\n",
       "5     Avpr2  HP_0001742  0.564504\n",
       "6     Chrm3  HP_0001742  0.565059\n",
       "7     Trpm8  HP_0001742  0.564504\n",
       "8   Sigmar1  HP_0001742  0.606034\n",
       "9    Adra2a  HP_0001742  0.600616\n",
       "10   Avpr1a  HP_0001742  0.564504\n",
       "11    Adrb2  HP_0001742  0.540921\n",
       "12   Adra1b  HP_0001742  0.606796\n",
       "13   Adra1d  HP_0001742  0.607350\n",
       "14   Grin2a  HP_0001742  0.606034\n",
       "15   Adra2c  HP_0001742  0.600616\n",
       "16    Adrb3  HP_0001742  0.540921\n",
       "17     Hrh1  HP_0001742  0.606959\n",
       "18   Avpr1b  HP_0001742  0.564504\n",
       "19   Adra2b  HP_0001742  0.600616"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_ID = 'HP_0001742' #MONDO_0023369\n",
    "min_score = 0.4\n",
    "\n",
    "disease_list = [disease_ID] #disease list starts with the disease ID\n",
    "print('including descendants')\n",
    "descendant_list = diseases.filter(col(\"id\")==disease_ID).select('descendants',).first()[0] #list of descendants for specific disease id\n",
    "print('found', len(descendant_list), 'descendants')\n",
    "disease_list = disease_list + descendant_list #concatenate two lists\n",
    "disease_list\n",
    "\n",
    "gene_list_df = associationByOverallDirect.filter(col(\"diseaseId\").isin(disease_list)).filter(col(\"score\")>min_score).select(['targetId','diseaseId','score']).toPandas()\n",
    "\n",
    "ortholog_map = 'GWAS/ensembl_human_orthologues/Human_to_Mouse.txt'\n",
    "\n",
    "ortholog_map_df = pd.read_csv(ortholog_map)\n",
    "ortholog_map_df = ortholog_map_df.dropna()\n",
    "gene_list = list(gene_list_df['targetId'])\n",
    "input_gene_count = len(gene_list)\n",
    "ortholog_map_dict = dict(zip(ortholog_map_df.iloc[:,0], ortholog_map_df.iloc[:,3]))\n",
    "gene_list_df['targetId'] = gene_list_df['targetId'].map(ortholog_map_dict)\n",
    "\n",
    "#gene_list = [x for x in map(ortholog_map_dict.get, gene_list) if x is not None]\n",
    "gene_list_df = gene_list_df[gene_list_df['targetId'].notna()]\n",
    "print(len(gene_list_df), 'focal species genes mapped to', input_gene_count, 'human genes.')\n",
    "\n",
    "gene_list_df.sort_values('diseaseId').to_csv(disease_ID + '_genes_df.csv',index = False)\n",
    "gene_list_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8d35502c-ddfe-4cf2-b5f0-9e54c5814511",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "including descendants\n",
      "found 0 descendants\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Stage 120:=============>                                        (13 + 40) / 53]\r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1325 focal species genes mapped to 1369 human genes.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetId</th>\n",
       "      <th>variantRsId</th>\n",
       "      <th>studyId</th>\n",
       "      <th>score</th>\n",
       "      <th>diseaseFromSourceMappedId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Clic4</td>\n",
       "      <td>rs6686889</td>\n",
       "      <td>GCST004280</td>\n",
       "      <td>0.699869</td>\n",
       "      <td>EFO_0006336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Clic4</td>\n",
       "      <td>rs4601530</td>\n",
       "      <td>GCST90000059</td>\n",
       "      <td>0.770126</td>\n",
       "      <td>EFO_0006336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Clic4</td>\n",
       "      <td>rs5773100</td>\n",
       "      <td>NEALE2_4079_raw</td>\n",
       "      <td>0.747710</td>\n",
       "      <td>EFO_0006336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Clic4</td>\n",
       "      <td>rs4543739</td>\n",
       "      <td>GCST90018952</td>\n",
       "      <td>0.714260</td>\n",
       "      <td>EFO_0006336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Fes</td>\n",
       "      <td>rs2521501</td>\n",
       "      <td>GCST001228</td>\n",
       "      <td>0.751210</td>\n",
       "      <td>EFO_0006336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1364</th>\n",
       "      <td>Zbtb10</td>\n",
       "      <td>rs6990531</td>\n",
       "      <td>GCST90018952</td>\n",
       "      <td>0.771132</td>\n",
       "      <td>EFO_0006336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1365</th>\n",
       "      <td>Zbtb10</td>\n",
       "      <td>rs11993814</td>\n",
       "      <td>NEALE2_4079_raw</td>\n",
       "      <td>0.838807</td>\n",
       "      <td>EFO_0006336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1366</th>\n",
       "      <td>Zbtb10</td>\n",
       "      <td>rs150539196</td>\n",
       "      <td>GCST90025981</td>\n",
       "      <td>0.827740</td>\n",
       "      <td>EFO_0006336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1367</th>\n",
       "      <td>Col4a2</td>\n",
       "      <td>rs9559779</td>\n",
       "      <td>NEALE2_4079_raw</td>\n",
       "      <td>0.842621</td>\n",
       "      <td>EFO_0006336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1368</th>\n",
       "      <td>Col4a2</td>\n",
       "      <td>rs9559779</td>\n",
       "      <td>GCST90018952</td>\n",
       "      <td>0.841282</td>\n",
       "      <td>EFO_0006336</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1325 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     targetId  variantRsId          studyId     score  \\\n",
       "0       Clic4    rs6686889       GCST004280  0.699869   \n",
       "1       Clic4    rs4601530     GCST90000059  0.770126   \n",
       "2       Clic4    rs5773100  NEALE2_4079_raw  0.747710   \n",
       "3       Clic4    rs4543739     GCST90018952  0.714260   \n",
       "4         Fes    rs2521501       GCST001228  0.751210   \n",
       "...       ...          ...              ...       ...   \n",
       "1364   Zbtb10    rs6990531     GCST90018952  0.771132   \n",
       "1365   Zbtb10   rs11993814  NEALE2_4079_raw  0.838807   \n",
       "1366   Zbtb10  rs150539196     GCST90025981  0.827740   \n",
       "1367   Col4a2    rs9559779  NEALE2_4079_raw  0.842621   \n",
       "1368   Col4a2    rs9559779     GCST90018952  0.841282   \n",
       "\n",
       "     diseaseFromSourceMappedId  \n",
       "0                  EFO_0006336  \n",
       "1                  EFO_0006336  \n",
       "2                  EFO_0006336  \n",
       "3                  EFO_0006336  \n",
       "4                  EFO_0006336  \n",
       "...                        ...  \n",
       "1364               EFO_0006336  \n",
       "1365               EFO_0006336  \n",
       "1366               EFO_0006336  \n",
       "1367               EFO_0006336  \n",
       "1368               EFO_0006336  \n",
       "\n",
       "[1325 rows x 5 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_ID = 'EFO_0006336' #MONDO_0023369\n",
    "min_score = 0.6\n",
    "\n",
    "disease_list = [disease_ID] #disease list starts with the disease ID\n",
    "print('including descendants')\n",
    "descendant_list = diseases.filter(col(\"id\")==disease_ID).select('descendants',).first()[0] #list of descendants for specific disease id\n",
    "print('found', len(descendant_list), 'descendants')\n",
    "disease_list = disease_list + descendant_list #concatenate two lists\n",
    "gene_list_df = evidence_ot_genetics_portal.filter(col(\"diseaseFromSourceMappedId\").isin(disease_list)).filter(col(\"score\")>min_score).select(['targetId','variantRsId','studyId','score','diseaseFromSourceMappedId']).toPandas()\n",
    "ortholog_map = 'GWAS/ensembl_human_orthologues/Human_to_Mouse.txt'\n",
    "\n",
    "ortholog_map_df = pd.read_csv(ortholog_map)\n",
    "ortholog_map_df = ortholog_map_df.dropna()\n",
    "gene_list = list(gene_list_df['targetId'])\n",
    "input_gene_count = len(gene_list)\n",
    "ortholog_map_dict = dict(zip(ortholog_map_df.iloc[:,0], ortholog_map_df.iloc[:,3]))\n",
    "gene_list_df['targetId'] = gene_list_df['targetId'].map(ortholog_map_dict)\n",
    "\n",
    "#gene_list = [x for x in map(ortholog_map_dict.get, gene_list) if x is not None]\n",
    "gene_list_df = gene_list_df[gene_list_df['targetId'].notna()]\n",
    "print(len(gene_list_df), 'focal species genes mapped to', input_gene_count, 'human genes.')\n",
    "gene_list_df.sort_values('diseaseFromSourceMappedId').to_csv(disease_ID+'_GWAS_genes_df.csv',index = False)\n",
    "gene_list_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2e651f99-c931-451d-b93b-f6c6700b5892",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "including descendants\n",
      "found 660 descendants\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                \r"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2473 focal species genes mapped to 2602 human genes.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>targetId</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Myh7</td>\n",
       "      <td>EFO_0002945</td>\n",
       "      <td>0.530581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Tpm1</td>\n",
       "      <td>EFO_0002945</td>\n",
       "      <td>0.370596</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Lmna</td>\n",
       "      <td>EFO_0002945</td>\n",
       "      <td>0.688291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Acta1</td>\n",
       "      <td>MONDO_0100084</td>\n",
       "      <td>0.539913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Get3</td>\n",
       "      <td>MONDO_0859358</td>\n",
       "      <td>0.452735</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2597</th>\n",
       "      <td>Sdha</td>\n",
       "      <td>MONDO_0100294</td>\n",
       "      <td>0.801257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2598</th>\n",
       "      <td>Sdhb</td>\n",
       "      <td>MONDO_0100294</td>\n",
       "      <td>0.398199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2599</th>\n",
       "      <td>Sdhd</td>\n",
       "      <td>MONDO_0100294</td>\n",
       "      <td>0.687092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2600</th>\n",
       "      <td>Sdhaf1</td>\n",
       "      <td>MONDO_0100294</td>\n",
       "      <td>0.743727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2601</th>\n",
       "      <td>Nr3c1</td>\n",
       "      <td>EFO_0003063</td>\n",
       "      <td>0.503594</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2473 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     targetId      diseaseId     score\n",
       "0        Myh7    EFO_0002945  0.530581\n",
       "1        Tpm1    EFO_0002945  0.370596\n",
       "2        Lmna    EFO_0002945  0.688291\n",
       "3       Acta1  MONDO_0100084  0.539913\n",
       "4        Get3  MONDO_0859358  0.452735\n",
       "...       ...            ...       ...\n",
       "2597     Sdha  MONDO_0100294  0.801257\n",
       "2598     Sdhb  MONDO_0100294  0.398199\n",
       "2599     Sdhd  MONDO_0100294  0.687092\n",
       "2600   Sdhaf1  MONDO_0100294  0.743727\n",
       "2601    Nr3c1    EFO_0003063  0.503594\n",
       "\n",
       "[2473 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "disease_ID = 'MONDO_0003939' #MONDO_0023369\n",
    "min_score = 0.2\n",
    "\n",
    "disease_list = [disease_ID] #disease list starts with the disease ID\n",
    "print('including descendants')\n",
    "descendant_list = diseases.filter(col(\"id\")==disease_ID).select('descendants',).first()[0] #list of descendants for specific disease id\n",
    "print('found', len(descendant_list), 'descendants')\n",
    "disease_list = disease_list + descendant_list #concatenate two lists\n",
    "disease_list\n",
    "\n",
    "gene_list_df = associationByOverallDirect.filter(col(\"diseaseId\").isin(disease_list)).filter(col(\"score\")>min_score).select(['targetId','diseaseId','score']).toPandas()\n",
    "\n",
    "ortholog_map = 'GWAS/ensembl_human_orthologues/Human_to_Mouse.txt'\n",
    "\n",
    "ortholog_map_df = pd.read_csv(ortholog_map)\n",
    "ortholog_map_df = ortholog_map_df.dropna()\n",
    "gene_list = list(gene_list_df['targetId'])\n",
    "input_gene_count = len(gene_list)\n",
    "ortholog_map_dict = dict(zip(ortholog_map_df.iloc[:,0], ortholog_map_df.iloc[:,3]))\n",
    "gene_list_df['targetId'] = gene_list_df['targetId'].map(ortholog_map_dict)\n",
    "\n",
    "#gene_list = [x for x in map(ortholog_map_dict.get, gene_list) if x is not None]\n",
    "gene_list_df = gene_list_df[gene_list_df['targetId'].notna()]\n",
    "print(len(gene_list_df), 'focal species genes mapped to', input_gene_count, 'human genes.')\n",
    "gene_list_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "f12eedb4-163f-4330-adcd-9bb4503af0f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list_df.sort_values('diseaseId').to_csv(disease_ID + '_genes_df.csv',index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43744bff-dff7-4a00-b1af-c322253b1357",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gwas_query",
   "language": "python",
   "name": "gwas_query"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
